File: 04 Grid Search/grid.py
# !/usr/bin/env python
# -*- coding: utf-8 -*-

import numpy as np 
from blind import *

def GridSearch(step, lower, upper, Fx, type="min"):
	D=len(step)	#dimension 
	dominio = [None] * D 	#dominio de valores
	#L = [False]* D 	#vector auxiliar
	for i in range(D):
		dominio[i]= np.arange(lower[i], upper[i], step[i])
		#L[i] = len(dominio[i])
	#LS = np.prod(L)
	#s = np.ndarray(shape=(LS, D)) #espacio de busqueda
	#k=0
	#for i in range(D,0,-1):
	#	for j in range(LS):
	#		for k in range(len(dominio[i])):
	#			E = 1 if i==0 else E*L[i-1]
	#			#print(len(dominio[i]))
	#			#print(len(s[i]))
	#			s[j][i] = dominio[i][k]
	#			#k = 0 if (k+1)==len(dominio[i]) else k+1
	#s = np.matrix(s)
	#print(s)
	#print(dominio)
	#print(len(s))
	#return fsearch(Fx, s, type)	# mejor solucion 
	return dfsearch(dominio, Fx, type=type)
var = 5
upper = [1000]*var
lower = [1]*var
step = [100]*var 

def f(x):
	return pow(x,2)

valor= GridSearch(step, lower, upper, f, "max")
print(valor)

File: 04 Grid Search/grid.R
### grid.R file ###

source("C:/Users/adrsosa/OneDrive - Universidad Veracruzana/Programación Estadistica/03 Blind Search/fsearch.R", encoding = "utf-8")

#		Metodo estandar de búsqueda en maya (búsqueda ciega amplitud)
#		step	-	vector con el tamaño del paso para cada dimensión D
#		lower	-	vector con el valor mas bajo de cada dimensión
#		upper	-	vector con el valor mas alto de cada dimensión
#		FUN		-	Función a evaluar
#		type	-	"min"o "max"
#		...		-	parametros extras para FUN

gsearch <- function(step,lower,upper,FUN,type="min",...)
{
  # dimensión
  D<-length(step)
  # valores de dominio
  domain<-vector("list",D)
  # vector auxiliar
  L <- vector(length=D)
  # llenamos domain con una secuencia d enumeros desde lower hasta upper en pasos de step
  # se actualiza el vector auxiliar con la longitud de cada arreglo de domain
  for(i in 1:D)
  { 
    domain[[i]]=seq(lower[i],upper[i],by=step[i])
    L[i]=length(domain[[i]])
  }
  # obtenemos el producto de L
  LS <- prod(L)
  # definimos el espacio de busqueda
  s <- matrix(ncol=D, nrow=LS)
  for (i in 1:D)
  {
    if (i==1) E <-1 else E <- E+L[i+1]			# se define cada cuanto se repite la serie 
    s[,i] <- rep(domain[[i]], length.out=LS, each=E)	# llena el espacio de busqueda con valores sucesivos definidos por E y domain
  }
  fsearch(sol=s, Fx=FUN,type=type,...)		#buscamos la mejor solución
}

#		Metodo estandar de búsqueda en maya (búsqueda ciega profundidad)
#		step	-	vector con el tamaño del paso para cada dimensión D
#		lower	-	vector con el valor mas bajo de cada dimensión
#		upper	-	vector con el valor mas alto de cada dimensión
#		FUN		-	Función a evaluar
#		type	-	"min"o "max"
#		...		-	parametros extras para FUN

fgsearch <- function(step,lower,upper,FUN,type="min",...)
{
	D = length(step)						# dimensiones
	dominio=vector("list",D)					# valores de dominio
	for(i in 1:D) dominio[[i]]=seq(lower[i],upper[i],by=step[i])		# llena el dominio de busqueda con valors secuenciales dentre upper y lower en steps

	dfsearch(dominio=dominio, Fx=FUN, type=type, ...) # solution
}